### DISCUSSION TITLE
Using TCP Serial Bridge?

### DISCUSSION THREAD

**Discussion by dachshund-digital** [arendst/Tasmota#12239] [Q&A] [answered]
I have connected a TAS TCP Serial Bridge to a Raspberry Pi, so that I can connect to the serial UART on the Pi via said TCP serial bridge.  This works, I see the Pi serial output, and I can interact with the connection using say telnet, or even PuTTY raw.  But when I type a command, and hit return, the command is echoed to screen.  This is a remote echo, not a local echo.  How do I turn off this remote echo (i.e. echo from the ESP device)?

What I expect is...

# ls -l
total 0

(For an empty directory)

But what I get is...
# ls -l
ls -l
total 0

Any inputted command, is echoed back to the active session, when I hit return.  As I understand it, this is some type of remote echo.  But I don't see any obvious way to disable it on the TAS side?

I could create an app or such to just ignore the extra echo, versus using PuTTy in raw mode or even telnet.  But that seems kind of cheesy.  Anyone have a better idea or suggestion?  Would like to use standard terminal apps if possible if there is some way to disable this remote echo?

**Comment by sfromis** [comment]
I'd expect that echo to be from your telnet client, with the option of controlling it there. Or maybe from the telnet server on the Pi. When using TCP Serial Bridge to talk to another Tasmota device via serial, I see no extra echo of the input I've typed, outside of the very explicit `CMD:` output from Tasmota, but that's not echo as such.

**Comment by s-hadinger** [comment, marked as answer]
Tasmota does not echo anything. It must come from elsewhere.

**Comment by dachshund-digital** [comment]
Weird, telnet and putty even if I disable local echo, I still get the entire line of text repeated after I hit enter.  I changed putty to a raw session, same result as well.  I guess I will have to walk through the ESP link code and see what they are doing for their emulation of the serial console.  It is structured much like TAS console... the command input is separate from the output, versus how telnet and putty do it more like a traditional terminal (interactive) session.  As an emergency console, it works as is, just a bit goofy.

Oh, like sfromis noted, at 115200 baud rate, I get buffer issues at times.  Going to set Pi UART to maybe 19200 or so, and test a bit more.  Even 9600 would be ok as an emergency console resource.

**Reply by sfromis** [reply]
It might be a question of your PI serial port being configured for full vs half duplex.

**Reply by dachshund-digital** [reply]
And that is a good suggestion, will check.  I connected termite direct to a USB to Serial adapter, and see the same behavior, which confirms it is coming from the Pi setup.  NOT that I did not already believe that was the case based on our discussion.  Just confirmation at this point.

**Reply by dachshund-digital** [reply]
It was not the duplex setting.  It was something in the Pi configuration.  I keep thinking that this entire issue was just odd, so I rebuilt the Pi SD card image from scratch.  The new image build is not generating the duplicate text on newline/enter return.  Which makes sense if one respect, because every tool I used showed the extra text, PuTTy, Termite, Telnet, etc.  I have more applications and configuration changes to do to the image to get it back to the previous one references to apps content, etc.  So I should be able to figure out what triggered the issue as some point.
